pkg/importer/spanner/sysl.arrai

```diff
@@ -0,0 +1,140 @@
+let entityPattern = \entity \model \mode
+    cond {
+        mode = 0 && entity.primary_key count > 1 && compareColumnOrder(entity):
+            $`primary_key="${//seq.join(', ', entity.primary_key orderby .)}"`,
+        mode = 1 && entity.cluster:
+            //seq.join('', entity.cluster >> $`interleave_in_parent="${.interleaved_in}", interleave_on_delete="${.on_delete}"`),
+        mode = 2 && entity.foreign_keys:
+            $`foreign_keys=[${//seq.join(', ',(entity.foreign_keys => \keys $`["constraint:${keys.constraint_name}` ++
+            $`","columns:${//seq.join(', ', (keys.foreign_keys => .attribute) orderby .)}"]`) orderby .)}]`,
+        let relatedIndices = ((model.indexes orderby .) where .@item.table_name = entity.name); mode = 3 && relatedIndices:
+            $`indexes=[${//seq.join(', ', ((model.indexes orderby .) where .@item.table_name = entity.name) >>
+            $`"name:${.table_name}` ++
+            $`","unique:${cond {(.unique = 'UNIQUE'):'true', _:'false'}}` ++
+            $`","null_filtered:${cond {(.nullfiltered = 'NULL_FILTERED'):'true', _:'false'}}` ++
+            $`","key_parts:${//seq.join(', ', .key_part orderby .)}` ++
+            cond {.storing_col: $`","storing:' ${//seq.join(', ',.storing_col)}`}  ++
+            cond {.interleaved_table: $`","interleave_in:' ${.interleaved_table}`})}"]`,
+    };
+
+# allEntityPatterns generates the pattern string for an entity
+let allEntityPatterns = \entity \model
+    $`${entityPatternDelimiter(entity, model, '[')}${cond {
+        hasEntityPattern(entity, model):
+        //seq.join(', ',[entityPattern(entity, model, 0), entityPattern(entity, model, 1),
+        entityPattern(entity, model, 2), entityPattern(entity, model, 3)] where .@item != '')
+    }}${entityPatternDelimiter(entity, model, ']')}`;
+
+# allAttributePatterns generates the pattern string for an entity attribute
+let allAttributePatterns = \entity \attr
+    $`${attributePatternDelimiter(entity,attr,'[')}${cond {
+        hasAttributePattern(entity,attr):
+        //seq.join(', ',[attributePattern(entity,attr,2), attributePattern(entity,attr,0), attributePattern(entity,attr,1),
+        attributePattern(entity,attr,3), attributePattern(entity,attr,4)] where .@item != '')
+    }}${attributePatternDelimiter(entity,attr,']')}`;
+
+# typeInfo generates the type info for an attribute
+let typeInfo = \entity \attr \type \isArray
+    let fkAttr = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_attribute orderby .) where .;
+    let fkTable = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_table orderby .) where .;
+    cond {
+        fkAttr | fkTable:  //seq.join('', (fkTable => //seq.join('', .) orderby .)) ++ '.'
+            ++ //seq.join('', (fkAttr => //seq.join('', .) orderby .)),
+        isArray: $`sequence of ${type}`,
+        _: type,
+    };
```

Comment:
There are a lot of levels of stuff going on here... hard to follow.

I think this does the same thing, but makes it clearer what the logic is that is acting on the inputs.
```suggestion
let typeInfo = \entity \attr \type \isArray
    let attrFks = entity.foreign_keys => .foreign_keys where .attribute = attr.name;
    let fkAttr = attrFks => .reference_attribute where .;
    let fkTable = attrFks => .reference_table where .;
    # Joins a set of sets into a single alphabetically-ordered string.
    let flatJoin = \sofs $`${sofs => $`${. orderby .::}` orderby .::}`;
    cond {
        fkAttr | fkTable: $`${flatJoin(fkTable)}.${flatJoin(fkAttr)}`,
        isArray: $`sequence of ${type}`,
        _: type,
    };
```

Reply:
let flatJoin = \sofs $`${sofs => $`${. orderby .::}` orderby .::}`; breaks with a missing close. But all the { seems to be in place.